File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head==null||head.next==null){
            return null;
        }
        ListNode f = head;
        ListNode s = head;
        while (f != null && f.next != null) {
  
                f=f.next.next;
                s=s.next;
                if(f==s){
                    break;
                }
        }
        if(f == null|| f.next == null){
                return null;
        }
        ListNode temp=head;
        while(temp!=s){
            temp=temp.next;
            s=s.next;
        }
        return s;
        
    }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow=head
        fast=head
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next
            if slow is fast:
                return slow
        return None
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        temp=self.hasCycle(head)
        if not temp:
            return None
        c=head
        while(c is not temp):
            c=c.next
            temp=temp.next
        return c
        